disperser/dataapi/service_availability_handler.go

```diff
@@ -0,0 +1,168 @@
+package dataapi
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	mutex sync.Mutex
+)
+
+func (s *server) getServiceAvailability(ctx context.Context, services []string) ([]*ServiceAvailability, error) {
+
+	if services == nil {
+		return nil, fmt.Errorf("services cannot be nil")
+	}
+
+	availabilityStatuses := make([]*ServiceAvailability, len(services))
+
+	for i, serviceName := range services {
+		pool, ok := s.getClientPool(serviceName)
+		if !ok {
+			return nil, fmt.Errorf("Invalid ServiceName: %s", serviceName)
+		}
+		conn, err := getClientConn(pool)
+		if err != nil {
+			return nil, fmt.Errorf("Error getting client connection: %v", err)
+		}
+		var availabilityStatus *ServiceAvailability
+		client := grpc_health_v1.NewHealthClient(conn)
+		response, err := client.Check(ctx, &grpc_health_v1.HealthCheckRequest{})
+
+		if err != nil {
+			availabilityStatus = &ServiceAvailability{
+				ServiceName:   serviceName,
+				ServiceStatus: grpc_health_v1.HealthCheckResponse_NOT_SERVING.String(),
+			}
+			availabilityStatuses[i] = availabilityStatus
+		} else {
+
+			availabilityStatus = &ServiceAvailability{
+				ServiceName:   serviceName,
+				ServiceStatus: response.Status.String(),
+			}
+			availabilityStatuses[i] = availabilityStatus
+		}
+
+		if availabilityStatuses[i].ServiceStatus == grpc_health_v1.HealthCheckResponse_SERVING.String() {
+			pool.clients <- conn
+		} else {
+			conn.Close()
+		}
+	}
+	return availabilityStatuses, nil
+}
+
+// Initializes the client pools for the server
+func (s *server) InitGRPCClientPools(poolSize int) error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	var err error
+	if s.clientPools == nil {
+		s.clientPools = make(map[string]*ClientPool)
+	}
+
+	// Register Disperser and Churner client pools
+	// These are the only public services that the dataapi server will connect to
+	s.clientPools["Disperser"], err = newClientPool(poolSize, s.disperserHostName)
+	if err != nil {
+		return err
+	}
+	s.clientPools["Churner"], err = newClientPool(poolSize, s.churnerHostName)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// newClientPool creates a client pool with prewarmed connections
+func newClientPool(size int, serverAddr string) (*ClientPool, error) {
```

Comment:
Why does it need a pool when it's queried once per minute by a single client?

Reply:
What will be resulting RPS?

Reply:
Feels like this is something that we should add after observing the system instead of preemptively optimizing for some future scenario. 

If it's already working, I would leave it.